### DISCUSSION TITLE
gs1:     { hash: 'gs1pk', sort: 'gs1sk', project: ['gs1pk', 'gs1sk', 'data'] },

### DISCUSSION THREAD

**Discussion by jonathanalberghini** [sensedeep/dynamodb-onetable#75] [Q&A] [answered]
I do not understand what the 'data' means?

**Comment by mobsense** [comment, marked as answer]
Those are attributes to project to the Index.  That means that any model that has a data attribute will be projected to the index.

**Reply by jonathanalberghini** [reply]
But when I look at the schemas I do not see any attributes called data. Am I missing something? Sorry I am new to onetable.
Thanks for answering.

**Reply by jonathanalberghini** [reply]
I was looking at this sample. https://github.com/sensedeep/dynamodb-onetable/blob/main/samples/overview/src/schema.js

**Reply by mobsense** [reply]
Cut and paste in one sample (overview). That is a mistake.

In the `packed` sample, you'll see:

```
                address:     { type: String, map: 'data.address' },
                city:        { type: String, map: 'data.city' },
                zip:         { type: String, map: 'data.zip' },
```

Note: the attributes in the indexes are ATTRIBUTES not properties. Address, city and zip are properties that are mapped into the `data` attribute.

**Reply by jonathanalberghini** [reply]
Is it the inclue projection type ? does onetable wrap objects in data field?
            address:    { type: Object, default: {}, schema: {
                street: { type: String, /* map: 'data.street' */ },
                city:   { type: String, /* map: 'data.city' */ },
                zip:    { type: String, /* map: 'data.zip' */ },
            } },

ProjectionType
The set of attributes that are projected into the index:

KEYS_ONLY - Only the index and primary keys are projected into the index.

INCLUDE - In addition to the attributes described in KEYS_ONLY, the secondary index will include other non-key attributes that you specify.

ALL - All of the table attributes are projected into the index.

**Reply by mobsense** [reply]
In that sample, the data attribute is projected as an object/map.  The street, city and zip properties are then stored as properties in the data attribute.  When you use project: [....], we use ProjectType == INCLUDE.

**Reply by jonathanalberghini** [reply]
ok Thanks?

**Reply by mobsense** [reply]
This is a really cool feature. You can project just the keys and a data attribute, but then you can modify your schema after creating the index to store any model properties under the `data` attribute. This uncouples the schema design from the index structure.

**Comment by mobsense** [comment]
I'll move this to the Discussions

**Comment by jonathanalberghini** [comment]
That is really cool so if I'm filtering by any of the data that I put into the index or I project into the index it won't do a scan because it has that data in the index is that the purpose of this?


Jon Alberghini
Director of Technology
***@***.***
o. 802-323-4558 | c. 802-323-4558 | www.legacyresearch.com<http://www.legacyresearch.com>

________________________________

**Reply by mobsense** [reply]
Not really.  The purpose is by projecting multiple properties into the index, you can access that data from the index in a single read. Normally, you'd have to project ALL attributes which drives up your data storage costs. With this, you decide on a per-model basis what gets projected into the index via a packed attribute like data.

**Reply by jonathanalberghini** [reply]
So if i projected say the field price to the GSI and I queries for invoices for the user with price > 100 Would that scan or query?

**Reply by mobsense** [reply]
First, a scan or query is always explicit by what API you call. ie. query / queryItems or scan / scanItems. You are never surprised.

You can query on the GSI using the PK and SK. You can then filter on the data attribute.  So you give property values for PK/SK and you can use a `where` expression (samples in the README) for constructing the filter expression.

**Comment by jonathanalberghini** [comment]
This is an awesome library!!! Do you have a sample of local indexes?

**Reply by mobsense** [reply]
No sample, but they are supported. Just set the hash value of the local index to the primary hash value.

**Comment by jonathanalberghini** [comment]
Thank you so much


Jon Alberghini
Director of Technology
***@***.***
o. 802-323-4558 | c. 802-323-4558 | www.legacyresearch.com<http://www.legacyresearch.com>

________________________________
